build_runner/bin/graph_inspector.dart

```diff
@@ -0,0 +1,203 @@
+
+  stdout.writeln('Ready, please type in a command:');
+
+  while (true) {
+    stdout.writeln('');
+    stdout.write('> ');
+    var nextCommand = stdin.readLineSync();
+    stdout.writeln('');
+    await commandRunner.run(nextCommand.split(' '));
+  }
+}
+
+class InspectNodeCommand extends Command {
+  @override
+  String get name => 'inspect';
+
+  @override
+  String get description =>
+      'Lists all the information about an asset using a relative or package: uri';
+
+  @override
+  String get invocation => '${super.invocation} <dart-uri>';
+
+  InspectNodeCommand() {
+    argParser.addFlag('verbose', abbr: 'v');
+  }
+
+  @override
+  run() {
+    var stringUris = argResults.rest;
+    if (stringUris.isEmpty) {
+      stderr.writeln('Expected at least one uri for a node to inspect.');
+    }
+    for (var stringUri in stringUris) {
+      var id = _getIdFromStringUri(stringUri);
+      if (id == null) {
+        continue;
+      }
+      var node = assetGraph.get(id);
+      if (node == null) {
+        stderr.writeln('Unable to find an asset node for $stringUri.');
+        continue;
+      }
+
+      var description = new StringBuffer()
+        ..writeln('Asset: $stringUri')
+        ..writeln('  type: ${node.runtimeType}');
+
+      if (node is GeneratedAssetNode) {
+        description.writeln('  wasOutput: ${node.wasOutput}');
+      }
+
+      _printAsset(AssetId asset) =>
+          _listAsset(asset, description, indentation: '    ');
+
+      if (argResults['verbose'] == true) {
+        description.writeln('  primary outputs:');
+        node.primaryOutputs.forEach(_printAsset);
+
+        description.writeln('  secondary outputs:');
+        node.outputs.difference(node.primaryOutputs).forEach(_printAsset);
+
+        if (node is GeneratedAssetNode) {
+          description.writeln('  inputs:');
+          var inputs = assetGraph.allNodes
+              .where((n) => n.outputs.contains(node.id))
+              .map((n) => n.id);
+          inputs.forEach(_printAsset);
+        }
+      }
+
+      stdout.write(description);
+    }
+  }
+}
+
+class GraphCommand extends Command {
+  @override
+  String get name => 'graph';
+
+  @override
+  String get description => 'Lists all the nodes in the graph.';
+
+  @override
+  String get invocation => '${super.invocation} <dart-uri>';
+
+  GraphCommand() {
+    argParser.addFlag('generated',
+        abbr: 'g', help: 'Show only generated assets.', defaultsTo: false);
+    argParser.addFlag('original',
+        abbr: 'o',
+        help: 'Show only original source assets.',
+        defaultsTo: false);
+    argParser.addOption('package',
+        abbr: 'p', help: 'Filters nodes to a certain package');
+    argParser.addOption('pattern',
+        abbr: 'm', help: 'glob pattern for path matching');
+  }
+
+  @override
+  run() {
+    var showGenerated = argResults['generated'] as bool;
+    var showSources = argResults['original'] as bool;
+    Iterable<AssetNode> nodes;
```

Comment:
[optional] Do you ever use anything from the node other than `id`? If not we could make this an `Iterable<AssetId>` and some of the lines below could be shorted to `assetGraph.sources` and `assetGraph.outputs`.

Reply:
done